PSM Data Feed Addition

Michael7371 opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description

This PR introduces support for PSM message processing to the ODE. Support for both required and optional fields is present, however I do not have a sample message that covers any of the optional fields for testing. All required fields have been tested with sample PSM messages generated from the NJDOT deployment.

This pipeline takes in PSM messages on the new UDP port (44940) and outputs processed JSON PSM messages on the topic: topic.OdePsmJson. As with other data feeds, multiple other topics are used for internal processing/management: topic.OdeRawEncodedPSMJson and topic.OdePsmTxPojo.

## Related Issue

There are no open GitHub issues related to this PR.

## Motivation and Context

Required for processing Personal Safety Messages through the ODE. 

## How Has This Been Tested?

I followed the convention from previous data feeds for unit testing and verified message processing with sample messages from the NJDOT deployment.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ x ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ x ] I have added any new packages to the sonar-scanner.properties file
- [ x ] My change requires a change to the documentation.
- [ x ] I have updated the documentation accordingly.
- [ x ] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/bugfix/Pull_request_template/docs/contributing_guide.md) 
- [ x ] I have added tests to cover my changes.
- [ x ] All new and existing tests passed.

drewjj reviewed (changes requested):
Everything checks out, runs well and tests well. I think this is ready to go but do you mind editing this schema documentation to include PSM? Link to the schema you have like the others do. It is an easier, more human, way to find the schemas.

Michael7371 commented:
> Everything checks out, runs well and tests well. I think this is ready to go but do you mind editing this schema documentation to include PSM? Link to the schema you have like the others do. It is an easier, more human, way to find the schemas.
> 
> https://github.com/Michael7371/jpo-ode/tree/psm-data-feed/docs/schemas

Gotcha, that should be updated now.

drewjj reviewed (approved):
Great, Looks good!

dmccoystephenson reviewed (changes requested):
With the addition of a new message type, the following should be done:
- The [ODE Data Flow Overview diagram](https://github.com/CDOT-CV/jpo-ode/blob/master/docs/data-flow-diagrams/ODE%20Data%20Flow%20Overview.drawio.png) should be updated to include the path that PSMs take.
- The existing overview data flow explanations should be updated as necessary in the README in the `docs/data-flow-diagrams` directory.
- A new subdirectory 'psm' should be added to the `docs/data-flow-diagrams` directory.
- In that new subdirectory a new diagram for the data flow of PSMs should be created.
- A new data flow explanation for PSMs should be added to the README in the `docs/data-flow-diagrams` directory.

Michael7371 commented:
@dmccoystephenson I pushed an update that addresses most of your comments. I didn't see what needed to be added to the ODE Data Flow Diagram, so I left that unchanged.

dmccoystephenson commented on jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java:
```diff
@@ -0,0 +1,43 @@
+public class OdePsmMetadata extends OdeLogMetadata {
+
+	/**
+	 * 
```
This comment seems unnecessary

dmccoystephenson commented on jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmMetadata.java:
```diff
@@ -0,0 +1,43 @@
+		super(payload);
+	}
+
+	public OdePsmMetadata(OdeMsgPayload payload, SerialId serialId, String receivedAt) {
+
+	}
```
Is this constructor necessary?

Michael7371 replied:
No, just came over from the BSM implementation (not sure it's useful over there either).

Michael7371 replied:
I'll remove it

dmccoystephenson commented on jpo-ode-core/src/main/java/us/dot/its/jpo/ode/model/OdePsmPayload.java:
```diff
@@ -0,0 +1,32 @@
+
+public class OdePsmPayload extends OdeMsgPayload {
+
+	   /**
+	 * 
+	 */
```
Same here, this comment can probably be removed

dmccoystephenson commented on jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PSM.java:
```diff
@@ -0,0 +1,240 @@
+
+public class J2735PSM extends Asn1Object {
+
+	/**
+	 * 
+	 */
```
This comment should be removed or populated

dmccoystephenson commented on jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PersonalDeviceUsageState.java:
```diff
@@ -0,0 +1,9 @@
+package us.dot.its.jpo.ode.plugin.j2735;
+
+public enum J2735PersonalDeviceUsageState {
+    unavailable,
+    aPEDESTRIAN,
+    aPEDALCYCLIST,
+    aPUBLICSAFETYWORKER,
+    anANIMAL
```
Are these the correct values for this enum? Looks like the same ones as the J2735PersonalDeviceUserType enum.

Michael7371 replied:
Good catch yeah I'll fix that.

dmccoystephenson replied:
Thanks!

dmccoystephenson commented on jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/J2735PropelledInformation.java:
```diff
@@ -0,0 +1,40 @@
+
+public class J2735PropelledInformation extends Asn1Object  {
+	
+    /**
+	 * 
+	 */
```
This comment can be removed

dmccoystephenson commented on jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/OdePsmDataCreatorHelper.java:
```diff
@@ -0,0 +1,59 @@
+
+public class OdePsmDataCreatorHelper {
+
+	public OdePsmDataCreatorHelper() {
+	}
```
Is this constructor necessary?

Michael7371 replied:
Honestly, I am not too sure. I just followed the convention with all the other Ode{dataflow name}DataCreatorHelper.java files.

dmccoystephenson reviewed (changes requested):
Looks good, just a few comments!

Also, the architecture diagram should be updated to include PSMs as input:
https://github.com/CDOT-CV/jpo-ode/blob/master/docs/images/readme/figure1.png

Michael7371 commented:
> Looks good, just a few comments! All tests are passing.
> 
> One thing: The architecture diagram should be updated to include PSMs as input: https://github.com/CDOT-CV/jpo-ode/blob/master/docs/images/readme/figure1.png

Do you know where the source file that generated that figure is stored?

dmccoystephenson reviewed (approved):
This looks good to me!
